File: system/multitier/backend-typescript/src/api/controller/order.controller.ts
import { Body, Controller, Get, Param, Post, Query, Res } from '@nestjs/common';
import type { Response } from 'express';
import { OrderService } from '../../core/services/order.service';
import { PlaceOrderRequest } from '../../core/dtos/place-order-request.dto';

@Controller('api/orders')
export class OrderController {
  constructor(private readonly orderService: OrderService) {}

  @Get()
  async browseOrderHistory(@Query('orderNumber') orderNumber?: string) {
    return this.orderService.browseOrderHistory(orderNumber);
  }

  @Post()
  async placeOrder(@Body() request: PlaceOrderRequest, @Res() res: Response) {
    const response = await this.orderService.placeOrder(request);
    res
      .status(201)
      .header('Location', `/api/orders/${response.orderNumber}`)
      .json(response);
  }

  @Get(':orderNumber')
  async getOrder(@Param('orderNumber') orderNumber: string) {
    return this.orderService.getOrder(orderNumber);
  }

  @Post(':orderNumber/cancel')
  async cancelOrder(
    @Param('orderNumber') orderNumber: string,
    @Res() res: Response,
  ) {
    await this.orderService.cancelOrder(orderNumber);
    res.status(204).send();
  }

  @Post(':orderNumber/deliver')
  async deliverOrder(
    @Param('orderNumber') orderNumber: string,
    @Res() res: Response,
  ) {
    await this.orderService.deliverOrder(orderNumber);
    res.status(204).send();
  }
}

File: system/multitier/backend-dotnet/Controllers/OrderController.cs
using Microsoft.AspNetCore.Mvc;
using MyCompany.MyShop.Backend.Core.Dtos;
using MyCompany.MyShop.Backend.Core.Services;

namespace MyCompany.MyShop.Backend.Controllers;

[ApiController]
[Route("api/orders")]
public class OrderController : ControllerBase
{
    private readonly OrderService _orderService;

    public OrderController(OrderService orderService)
    {
        _orderService = orderService;
    }

    [HttpGet]
    public async Task<IActionResult> BrowseOrderHistory([FromQuery] string? orderNumber)
    {
        var response = await _orderService.BrowseOrderHistoryAsync(orderNumber);
        return Ok(response);
    }

    [HttpPost]
    public async Task<IActionResult> PlaceOrder([FromBody] PlaceOrderRequest request)
    {
        var response = await _orderService.PlaceOrderAsync(request);
        return Created($"/api/orders/{response.OrderNumber}", response);
    }

    [HttpGet("{orderNumber}")]
    public async Task<IActionResult> GetOrder(string orderNumber)
    {
        var response = await _orderService.GetOrderAsync(orderNumber);
        return Ok(response);
    }

    [HttpPost("{orderNumber}/cancel")]
    public async Task<IActionResult> CancelOrder(string orderNumber)
    {
        await _orderService.CancelOrderAsync(orderNumber);
        return NoContent();
    }

    [HttpPost("{orderNumber}/deliver")]
    public async Task<IActionResult> DeliverOrder(string orderNumber)
    {
        await _orderService.DeliverOrderAsync(orderNumber);
        return NoContent();
    }
}
